Remove liquid from the world when filling container items.

4Denthusiast opened the pull request:
When filling fluid containers (like the cup and bucket) from water blocks, the amount of water in the block is depleted, and if there isn't enough water in the block to fill the container, it's only partially filled.

This adds a dependency on FlowingLiquids, as that contains the definition of liquid level in blocks. As a side-effect, this enables FlowingLiquids in Josharias Survival, and there's a bit of a problem with the interaction of FlowingLiquids and Caves (as the latter generates air-filled caves underwater, causing large amounts of liquid motion immediately after world generation).

I couldn't find an official definition of what the units of volume in this module are. I set the volume of liquid per full liquid block at 1000 (so the units are litres, assuming one block is 1m). This means that the cup and bucket from ManualLabor hold 0.3 and 0.5 blocks respectively. This seems like a sensible amount for the bucket to hold but an absurdly large amount for the cup, but that's more a product of the ratio that ManualLabor sets between them.

4Denthusiast committed:
Remove liquid from the world when filling container items.

4Denthusiast commented:
With this change, it would probably make sense to alter Thirst so that you can't drink more than you need. Otherwise you could get rid of large amounts of unwanted water by drinking it.

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -52,6 +56,24 @@
+    
+    @In
+    private ExtraBlockDataManager extraDataManager;
+    private int flowIx;
```
`flowIx`- what is an `Ix`?

4Denthusiast replied:
Oh, it's "index". I called it the same thing in FlowingLiquids's `LiquidFlowSystem` because it's used much more often there. I guess it doesn't really matter here, if you think the longer name would be better.

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -52,6 +56,24 @@
+    private ExtraBlockDataManager extraDataManager;
+    private int flowIx;
+    
+    private static final float fluidPerBlock = 1000;
```
I think Checkstyle should complain about this not being capitalized as constant:
```suggestion
    private static final float FLUID_PER_BLOCK = 1000;
```

In addition, you nicely describe how you come to that number in the PR description - would you mind adding this as doc comment here? 🙃

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -52,6 +56,24 @@
+    @Override
+    public void initialise() {
+        air = blockManager.getBlock(BlockManager.AIR_ID);
+        flowIx = extraDataManager.getSlotNumber("flowingLiquids.flow");
```
Is this string constant exposed by the FlowingLiquids module in some way? If not, I think that would be a good addition.
I'm thinking of something like `FlowingLiquids.FLOW` or `FlowingLiquids.EXTRA_DATA`.

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +118,44 @@
                 //TODO: replace with better fluid handling, maybe by new CoreFluids module
-                if (removedItem != null && block.isWater()) {
+                if (removedItem != null && worldProvider.getBlock(pos).isWater()) {
+                    byte liquidData = (byte) worldProvider.getExtraData(flowIx, pos);
```
I'd introduce a method for this, although this is already pretty concise. What do you think of static utility methods in FlowingLiquids itself to allow for simpler (and uniform) access to this data?

```java
public static byte getFlowAt(Vector3ic worldPosition, WorldProvider provider);
```

In turn, modules wouldn't even have to know the string key under which the data is stored as extra block data...

4Denthusiast replied:
I don't think that works as a static method: it needs access to the `ExtraBlockDataManager` to get the index. It would be possible to get the `LiquidFlowSystem` through the `ComponentSystemManager`, and have this as a non-static method in there (without the `WorldProvider` argument in that case, as it's already accessible). That wouldn't make the `FluidAuthoritySystem` so much neater though.

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +118,44 @@
+                    float blockAmount = LiquidData.getHeight(liquidData) * fluidPerBlock / LiquidData.MAX_HEIGHT;
+                    
+                    FluidContainerItemComponent fluidComponent = removedItem.getComponent(FluidContainerItemComponent.class);
+                    float totalAmount = blockAmount + fluidComponent.volume;
+                    if (totalAmount > fluidComponent.maxVolume) {
+                        blockAmount = totalAmount - fluidComponent.maxVolume;
+                        totalAmount = fluidComponent.maxVolume;
+                    } else {
+                        blockAmount = 0;
+                    }
```
Please create a method for this (IntelliJ's "extract to method" may help you here). The code itself is not complicated, but the body lambda expression is already quite lengthy and I'd like to slim this down.

4Denthusiast replied:
This block has 2 return values, so extracting it to a method would be excessively verbose. Also, why does it matter that this is inside of a lambda expression?

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +118,44 @@
         }
     }
+    
+    // Round to an integer in a way that has 0 error on average for any given argument.
```
Whut? Might be I'm already too tired to do helpful reviews, but could you try to explain in other words what we need this for/what problem this solves?

4Denthusiast replied:
I changed it to "Round randomly as either floor or ceiling in a way that has 0 error on average for any given argument.". Is that clearer? The idea is that fluids can't (on average) be created or destroyed by repeatedly picking them up and putting them down again (putting down not yet implemented), and also that if you pick from a full block with a very small container (less than 62.5l, the increment of liquid amount in blocks), it has a small chance of actually consuming any liquid.

skaldarnar reviewed (changes requested):
Integration of flowing liquids in JS would be cool!

How does it behave when taking water out of the ocean? I remember that at some point this would just propagate the water level change to infinity, basically bringing the game to it's knees. Did this behavior change in some way? 

Do we have some kind of "gameplay simiplification" in place that you cannot drown the ocean (should we)?

And, is there something we can do about the Caves? If enabling this causes some issues with performance and/or stability I'd be a bit cautios with merging this... overall **stabilization** and such 🙄

skaldarnar commented:
@ktksan FYI 😉

4Denthusiast committed:
Respond to review comments.

4Denthusiast commented:
> How does it behave when taking water out of the ocean? I remember that at some point this would just propagate the water level change to infinity, basically bringing the game to it's knees. Did this behavior change in some way?

When taking water from the ocean, it leaves a shallow dent. I'm planning on improving the rendering in FlowingLiquids so that dents in liquid surfaces are less noticable (displaying as slopes rather than steps), but haven't started that yet.

> Do we have some kind of "gameplay simiplification" in place that you cannot drown the ocean (should we)?

Currently there are no exceptions in FlowingLiquids to volume conservation (except possible one thing I'm not sure of involving liquid flowing into or out of unloaded chunks).  I'm fine with the oceans being finite. For the same reasons that you might consider making it impossible to drain the whole ocean, it's practically very difficult to do so anyway. The only problem I see here is that after liquid has been removed or added, the surface can be left slightly uneven. I've been working on trying to flatten the angle of repose, and as mentioned before, changing the rendering would also help.

> And, is there something we can do about the Caves? If enabling this causes some issues with performance and/or stability I'd be a bit cautios with merging this... overall **stabilization** and such roll_eyes

I have a plan to add another rasterizer to check for caves connected to the ocean and add water where necessary to avoid immediate flow. There would still be some problems with it, as neither the old cave algorithm nor my new one lend themselves to simplifications in connectivity calculations, but it would be much less severe than its current state.

Alternatively, Caves could just be modified so that there's always a solid barrier between caves and the ocean. This feels like an inelegant solution though, and underwater caves would be cool. It would be quicker and more reliable to implement though.

4Denthusiast committed:
Make FlowingLiquids dependency optional.

When picking up liquid with FL not enabled, the block is just entirely
removed or left in place, randomly with probabilities corresponding to
the amount of liquid removed.

4Denthusiast commented:
Now that the FlowingLiquids dependency is optional, it doesn't make FL a dependency of Josharias Survival, so the conflict with underwater caves in that, while still an issue, is no longer an issue with this PR.

If FL isn't enabled, picking up liquids just removed the whole block, with a probability proportional to the amount of fluid removed (otherwise it just leaves it be).

If anyone has suggestions for better ways of detecting whether FL is enabled than `componentSystemManager.get("FlowingLiquids:LiquidFlowSystem") != null`, that would be appreciated. This solution is not very forward-compatible.

4Denthusiast commented:
There's currently a bug that FlowingLiquids may take a while to notice that the liquid is missing, but the fix for that would be in FlowingLiquids, not here.
